### PR TITLE
Reuse Flatbuffer's ArrayBuffer during serialization.

### PR DESCRIPTION
**Pull request by ry**
This is a performance improvement to reduce the number of ArrayBuffer allocations.
```
> ./tools/throughput_benchmark.py out/before/deno 5000
105.897539854 seconds
> ./tools/throughput_benchmark.py out/after/deno 5000
100.990852118 seconds

> ./tools/http_benchmark.py out/before/deno
Requests/sec:  17464.64
> ./tools/http_benchmark.py out/after/deno
Requests/sec:  18223.10
```

### CODE REVIEW

File: js/flatbuffers.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  let bb = globalBuilder.dataBuffer();
+  // Only create a new backing ArrayBuffer if the previous one had grown very
+  // large in capacity. This should only happen rarely.
+  if (bb.capacity() > 1024) {
+    util.log(`realloc flatbuffer ArrayBuffer because it was ${bb.capacity()}`);
+    bb = ByteBuffer.allocate(1024);
+  }
+  // tslint:disable-next-line:no-any
+  const gb = globalBuilder as any;
+  gb["bb"] = bb;
```
Comment (piscisaureus) [inline js/flatbuffers.ts]: why are you using `gb["bb"]` vs `gb.bb` ?

File: js/flatbuffers.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+// This is a wrapper around the real Builder .
+// The purpose is to reuse a single ArrayBuffer for every message.
+// We can do this because the "control" messages sent to the privileged
+// side are guarenteed to be used during the call to libdeno.send().
+export function createBuilder(): Builder {
+  let bb = globalBuilder.dataBuffer();
+  // Only create a new backing ArrayBuffer if the previous one had grown very
+  // large in capacity. This should only happen rarely.
+  if (bb.capacity() > 1024) {
```
Comment (piscisaureus) [inline js/flatbuffers.ts]: 1kb is hardly what I would consider "very large". If we're able to reuse the buffer, why not allocate a fixed 64kb buffer?

File: js/flatbuffers.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+export type Offset = flatbuffers.Offset;
+export class ByteBuffer extends flatbuffers.ByteBuffer {}
+export class Builder extends flatbuffers.Builder {}
+
+// This is a wrapper around the real Builder .
+// The purpose is to reuse a single ArrayBuffer for every message.
+// We can do this because the "control" messages sent to the privileged
+// side are guarenteed to be used during the call to libdeno.send().
+export function createBuilder(): Builder {
+  let bb = globalBuilder.dataBuffer();
```
Comment (piscisaureus) [inline js/flatbuffers.ts]: You should somehow assert that the global buffer is not in use before reusing it.

File: js/flatbuffers.ts (hunk 2 above)
Comment (ry) [reply js/flatbuffers.ts]: These messages hardly never contain much data. (The one case I can think of is `readFile`, and I'd rather that be restructured to use the second "data" parameter to libdeno.send.) I'm just trying to get rid of all the tiny little "notification" messages.

File: js/flatbuffers.ts (hunk 3 above)
Comment (ry) [reply js/flatbuffers.ts]: Fixed.

File: js/flatbuffers.ts (hunk 2 above)
Comment (kitsonk) [reply js/flatbuffers.ts]: Feels like a "large message reallocation" might be a key metric to keep an eye on/externalise?

**Pull request merged**